util/topgen/secure_prng.py

```diff
@@ -0,0 +1,309 @@
+        N_iterations = requested_number_of_bits // self.blocklen
+        cipher = AES.new(self.Key.to_bytes(16, 'big'), AES.MODE_ECB)
+
+        for i in range(N_iterations):
+            # Step 4.1
+            # Because ctr_len = blocklen, steps 4.1.1 and 4.1.2 are not executed
+            self.V += 1
+            # Steps 4.2 and 4.3
+            pt = self.V
+            ct = cipher.encrypt(pt.to_bytes(16, 'big'))
+            # Step 4.3
+            temp = temp + ct
+
+        del (cipher)
+
+        # Step 5 is not needed because requested_number_of_bits % 128 == 0
+
+        # Step 6 additional_input = 0
+        self.CTR_DRBG_Update(0, 0)
+
+        # Step 7
+        self.reseed_counter += 1
+
+        # Step 8
+        return (temp)
+
+    def CTR_DRBG_Instantiate(self, entropy_input):
+        """ CTR_DRBG_Instantiate
+
+        CTR_DRBG Instantiate process, described in NIST SP 800-90A, page 52
+        Section 10.2.1.3.1 Instantiation When a Derivation Function is not used
+        entropy_input 256 bits
+        personalization_string is empty
+        """
+
+        # Steps 1, 2 and 3. Personalization string is empty
+        seed_material = entropy_input
+        # Step 4
+        self.Key = 0
+        # Step 5
+        self.V = 0
+        # Step 6. For convenience we split 256 bit seed_material into two 128b
+        # integers.
+        L = int.from_bytes(seed_material[0:16], 'big')
+        R = int.from_bytes(seed_material[16:32], 'big')
+        self.CTR_DRBG_Update(L, R)
+        # Step 7
+        self.reseed_counter = 1
+
+    def reseed(self, seed):
+        """Re-seed PRNG
+
+        Instantiates CTR_DRBG with a 256-bit entropy_input.
+        Empties returned_bits list in case that it contains residual numbers
+        from previous instantiation.
+        """
+        if seed == 0:
+            entropy_input = seed.to_bytes(32, 'big')
+            log.warning("WARNING: PRNG seeded with 0.")
+        # Check if the provided seed is at least 32 bytes long
+        elif sys.getsizeof(seed) < sys.getsizeof(1 << 255):
+            # Cyclically pad the seed to 256 bits and issue a warning
+            new_seed = seed
+            seed_bytes = _ceil(_log(seed + 1, 256))
+            while new_seed < (1 << 256):
+                new_seed <<= seed_bytes * 8
+                new_seed += seed
+            new_seed %= (1 << 256)
+            log.warning("Seed smaller than 256 bits. CTR_DRBG seeded with: " +
+                        hex(new_seed))
+            entropy_input = new_seed.to_bytes(32, 'big')
+        # Check if the seed is longer than 256 bits. Trim the excess bits and
+        # issue a warning if it is.
+        elif seed > (1 << 256):
+            new_seed = seed % (1 << 256)
+            log.warning("Seed longer than 256 bits. CTR_DRBG seeded with: " +
+                        hex(new_seed))
+            entropy_input = new_seed.to_bytes(32, 'big')
+        else:
+            entropy_input = seed.to_bytes(32, 'big')
```

Comment:
Is there any expected control flow where we reseed the PRNG with the wrong length of bits? I'd expect that in this case we'd exit with an error, not just a warning, since it's a security-critical operation. It's probably okay to take bytes here rather than an integer, which makes the length much easier to check.
```suggestion
    def reseed(self, seed):
        """Re-seed PRNG

        Instantiates CTR_DRBG with a 256-bit entropy_input.
        Empties returned_bits list in case that it contains residual numbers
        from previous instantiation.
        """
        if int.from_bytes(seed, byteorder='big') == 0:
            log.warning('WARNING: PRNG seeded with 0.')
        if len(seed)  != 32:
            log.error('Seed must be 32 bytes long.')
            sys.exit(1)
        entropy_input = seed
```

Reply:
> Is there any expected control flow where we reseed the PRNG with the wrong length of bits?

Unfortunately there is. Dvsim by default uses `--build-seed` of only 32 bits. Also, some people might run simulations with randomly typing seed parameters of arbitrary length.
I'm also not exactly sure how we choose the seed for the nightly regressions.
I guess that the correct thing to do would be to modify dvsim to check/expand the seed and to always make reseed calls with 256 bit. For the time being I've added a TO DO comment in the code and I'll open an issue to track this.